### PR TITLE
MDL

### PR DESCRIPTION
**Pull request by rohjha**

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -213,134 +207,146 @@ def main():
...
-            test_set = data_handler.make_data("test", None, test=True)
-            valid_set = data_handler.make_data("valid", None, test=True)
-        else:
-            train_set = data_handler.make_classifier_data("train", train_size)
-            test_set = data_handler.make_classifier_data("test", None, test=True)
-            valid_set = data_handler.make_classifier_data("valid", None, test=True)
+
+            mdls.append(test_loss)
+            print("Finished training on split {} / {}. Cost to trasmit the next split: {}.".format(i + 1, len(splits) - 1, test_loss))
+
```
Comment (cjlovering) [inline main.py:323]: We should compute data / code mdl here. As I understand it, the data cost is the train loss of a model trained on the whole train set (and the test set is the train set). The loss should be negative here:
```
model = mdl + train_loss
data = -train_loss
```

File: main.py [hunk 2]
```diff
@@ -213,134 +207,146 @@ def main():
...
+        
+        # How much did we overshoot by? We'll just take this from the longest dataset (TODO)
+        extra = np.sum(split_sizes) - len(train_set)
+        split_sizes[len(split_proportions) - 1] -= extra
+
+        splits = torch.utils.data.random_split(train_set, split_sizes.astype(int))
+        mdls = []
+
+        # Cost to transmit the first via a uniform code
+        mdls.append(split_sizes[0])
```
Comment (cjlovering) [inline main.py:281]: This looks correct: `log_2(num_classes = 2) = 1`.

File: main.py [hunk 3]
```diff
@@ -213,134 +207,146 @@ def main():
...
+        args.true_property,
+        args.noise,
+        counter_example_options,
+        args.hold_out,
+        args.experiment_id,
+        args.num_distractors,
+        args.num_unremovable_distractors,
+        args.initial_true_only_examples
+    )
+    if not args.train_classifier:
```
[the remark this answers is not in the document]
Comment (rohjha) [reply main.py:255]: Yes sounds good.

File: main.py (hunk 1 above)
Comment (rohjha) [reply main.py:323]: Will add this outside the loop. model = total_mdl - training_loss

File: main.py [hunk 4]
```diff
@@ -34,12 +34,11 @@
...
 
-device = torch.device("cpu")
-if "cuda" in args.device and torch.cuda.is_available():
-    device = torch.device(args.device)
+# NOTE: let's make sure it just always uses CUDA and fails otherwise
+device = torch.device("cuda")
 
 def make_dataloaders(train_set, test_set, valid_set):
-    num_workers = 1
+    num_workers = 0
```
Comment (cjlovering) [inline main.py:41]: We can just delete this line and use the default option -- its a bit confusing otherwise (I had to look at the doc to know why we would set it to 0).

File: main.py [hunk 5]
```diff
@@ -150,9 +146,8 @@ def train(train_dataloader, valid_dataloader, m, loss_function, optimizer):
             optimizer.step()
 
             # How many did the model get wrong?
-            for i in range(len(labels)):
-                if torch.argmax(probs[i]).item() != labels[i].item():
-                    train_wrong += 1
+            results = torch.argmax(probs, dim=1)
+            train_wrong += torch.sum(torch.abs(results - labels)).item() # This will be nonzero iff the model was wrong
```
Comment (cjlovering) [inline main.py:150]: Looks good! (If we had multiple labels it might miscount the number wrong, so we should keep it in mind if we ever change the labels for whatever reason.)
